### PR TITLE
Feat: Improve shape defaultProp rendering (#524)

### LINKED ISSUE #524: Improve printing of objects as `defaultProps`
Hi,

Before continuing, I would just like to say that I love react-styleguidist. it's an amazing product, very well put together. We evaluated many styleguide-libraries, but yours was by far the better in terms of producing something that was easy for the team to consume while working.

On to the issue at hand. My suggestion is quite simply to improve the rendering of `defaultProps` that are shapes. In the image below, you see that in this very simple example, it is breaking over several lines in a not very appealing way due to RSG falling back to rendering it as a string.

screen shot 2017-07-03 at 21 26 43

In this example, it's not **that** bad - but in another project I am working on, I would like to pass an object with several large strings and many properties, and it looks absolutely awful. The cell ends up taking too much space, and is generally unappealing - so we opted for not using `defaultProps` for our configuration objects for now. That's not to say that we don't **want** to.

What I propose is that the `defaultProps` field is rendered as the string `shape` or `Object` when an object is passed, and then on hover, focus or click, a popover is opened next to the string with the pretty printed corresponding `defaultProp` object.

Is this currently possible? Do you like the idea? I looked into the code, but it seems that the default `propsParser` does not return an immediately useful `defaultValue`. It would have to be parsed for it to be pretty printed. See below:

```js
{ type: { name: 'shape', value: {} },
  required: false,
  description: '',
  defaultValue: { value: '{\n    foo: \'bar\'\n}', computed: false },
  tags: {} }
```

### PR DESCRIPTION
**Pull request by esphen**
# What

This PR improves the printing of objects in `defaultProps`, which had some issues in the existing version. See #524 for further details.

# How

This PR removes the fallback of printing objects as strings. Instead, objects will be parsed, beautified, then finally displayed in a title. This mirrors the existing method used by `function` props.

The improved preview

Parsing is done with the [safe-eval](https://github.com/hacksparrow/safe-eval) library. Not sure if we actually need this library, though, as we are not in a node.js context here. Wanted to get some feedback first. We could use `eval` directly, or just render the unprettified source code representation in the title. Thoughts?

Fixes #524

### CODE REVIEW

**Comment by SaraVieira** [comment]
@esphen Will this only show on hover ?

**Comment by esphen** [comment]
@SaraVieira Yes, which brings me to an issue I have with the way these popovers are handled today. Seeing as you have to hover over the element to show the data, this excludes a potentially large audience from seeing this information. To quote the [HTML spec](https://www.w3.org/TR/html/dom.html#the-title-attribute):

> Relying on the title attribute is currently discouraged as many user agents do not expose the attribute in an accessible manner as required by this specification (e.g., requiring a pointing device such as a mouse to cause a tooltip to appear, which excludes keyboard-only users and touch-only users, such as anyone with a modern phone or tablet). 

I tried this in ChromeVox to verify, and was not able to read the text. So this solution is better for mouse users, but not not accessible, which is far from ideal. I would much prefer this text to be a `<button>` (which makes it focusable and tabbable), which shows a popover element on focus.

File: src/rsg-components/Props/PropsRenderer.js [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 import Group from 'react-group';
+import safeEval from 'safe-eval';
+import prettyFormat from 'pretty-format';
```
Comment (sapegin) [inline src/rsg-components/Props/PropsRenderer.js]: This looks like way too many new frontend dependencies for such feature. Have you compared the bundle size?

**Comment by sapegin** [comment]
> I tried this in ChromeVox to verify, and was not able to read the text. So this solution is better for mouse users, but not not accessible, which is far from ideal. I would much prefer this text to be a <button> (which makes it focusable and tabbable), which shows a popover element on focus.

Sounds like a good idea for a separate PR if we keep hover ;-)

**Comment by sapegin** [comment]
> Parsing is done with the safe-eval library. Not sure if we actually need this library, though, as we are not in a node.js context here.

I don’t think so since we’re rendering source code which I’d say is safe by default ;-)

> just render the unprettified source code representation in the title. Thoughts?

This or find some very simple library to do that. Pretty-print is very nice for Node but way too big for frontend. Something like [stringify-object](https://github.com/yeoman/stringify-object) maybe.

**Comment by esphen** [comment]
Thanks @sapegin. Based on your feedback I'll use `window.eval` for parsing and look into a smaller library for formatting.

**Comment by esphen** [comment]
@sapegin: I amended the commit with the requested changes. 

There were a lot of changes in the `package-lock.json` after I `rm package-lock.json && npm i` however. I wonder what that's about. Looks like it pruned all the `requires` fields, apart from that it looks pretty consistent with my expectations. I used npm@5.0.0 here.

**Comment by sapegin** [comment]
I would recommend to update to npm 5.2 — it’s much more stable but honestly I don’t really understand some changes it does to the lock file ;-|

**Comment by sapegin** [comment]
Oops, I forget that Sindre Sorhus doesn’t care about webpack users and his modules often breaks UglifyJS because they contains some ES6 code (see Travis CI logs) ;-(

Looks we have to choose yet another library ;-|

**Comment by esphen** [comment]
I found [javascript-stringify](https://github.com/blakeembrey/javascript-stringify), it looks pretty good. It has no dependencies. Pushed a new commit to the PR.

This is the result:

2017-07-14-194649_swaygrab

**Comment by n1313** [comment]
@esphen @sapegin do you think this change should be applied to the function popup as well? I would be nice to have these popups look the same.

**Comment by sapegin** [comment]
@n1313 I think he’s on Linux or something because it’s just a `title` ;-)

**Comment by n1313** [comment]
Titles on Linux look like *that*? o__o

**Comment by esphen** [comment]
@n1313 lol. Yes they do :)

**Comment by esphen** [comment]
@sapegin Travis seems to fail because it is now 5K over the recommended build size. What do you think?

**Comment by sapegin** [comment]
It’s fine to change it. This check is to prevent accidental adding of something huge to the bundle, like the whole Lodash ;-)

**Comment by esphen** [comment]
@sapegin: I pushed another commit which fixes the travis failure and fixes a merge conflict on `package-lock.json`. Are we good now?

**Comment by sapegin** [comment]
Looks good for me, thanks! Could you please add some tests too?

**Comment by esphen** [comment]
@sapegin Done, the new branches in the diff now have code coverage

**Comment by sapegin** [comment]
Merged, thanks!

**Pull request merged**